xdpderp/xdp.c

```diff
@@ -0,0 +1,513 @@
+	__u8 eth_tmp[ETH_ALEN];
+	__builtin_memcpy(eth_tmp, eth->h_source, ETH_ALEN);
+	__builtin_memcpy(eth->h_source, eth->h_dest, ETH_ALEN);
+	__builtin_memcpy(eth->h_dest, eth_tmp, ETH_ALEN);
+
+	// Flip ip header source and destination address.
+	if (is_ipv6) {
+		struct in6_addr ip_tmp = ip6->saddr;
+		ip6->saddr = ip6->daddr;
+		ip6->daddr = ip_tmp;
+	} else {
+		__be32 ip_tmp = ip->saddr;
+		ip->saddr = ip->daddr;
+		ip->daddr = ip_tmp;
+	}
+
+	// Flip udp header source and destination ports;
+	__be16 port_tmp = udp->source;
+	udp->source = udp->dest;
+	udp->dest = port_tmp;
+
+	// Trim packet to end of xor stun attribute.
+	if (is_ipv6) {
+		int shrink = (void *)(xor6 + 1) - data_end;
+		if (bpf_xdp_adjust_tail(ctx, shrink)) {
+			return end(ctx, XDP_ABORTED, is_ipv6);
+		}
+	} else {
+		int shrink = (void *)(xor + 1) - data_end;
+		if (bpf_xdp_adjust_tail(ctx, shrink)) {
+			return end(ctx, XDP_ABORTED, is_ipv6);
+		}
+	}
+
+	// Reset pointers post tail adjustment.
+	data_end = (void *)(long)ctx->data_end;
+	data = (void *)(long)ctx->data;
+	eth = data;
+	if ((void *)(eth + 1) > data_end) {
+		return end(ctx, XDP_ABORTED, is_ipv6);
+	}
+	if (is_ipv6) {
+		ip6 = (void *)(eth + 1);
+		if ((void *)(ip6 + 1) > data_end) {
+			return end(ctx, XDP_ABORTED, is_ipv6);
+		}
+	} else {
+		ip = (void *)(eth + 1);
+		if ((void *)(ip + 1) > data_end) {
+			return end(ctx, XDP_ABORTED, is_ipv6);
+		}
+	}
+
+	// Update ip header total length field and checksum.
+	__u32 cs = 0;
+	if (is_ipv6) {
+		if ((void *)(ip6 +1) > data_end) {
+			return end(ctx, XDP_ABORTED, is_ipv6);
+		}
+		__u16 payload_len = data_end - (void *)(ip6 + 1);
+		ip6->payload_len = bpf_htons(payload_len);
+	} else {
+		__u16 tot_len = data_end - (void *)ip;
+		ip->tot_len = bpf_htons(tot_len);
+		ip->check = 0;
+		cs = bpf_csum_diff(0, 0, (void *)ip, sizeof(*ip), cs);
+		ip->check = csum_fold_flip(cs);
+	}
+
+	// Avoid dynamic length math against the packet pointer, which is just a big
+	// verifier headache. Instead sizeof() all the things.
+	int to_csum_len = sizeof(*udp) + sizeof(*req) + sizeof(*sa);
+	// Update udp header length and checksum.
+	if (is_ipv6) {
+		to_csum_len += sizeof(*xor6);
+		udp = (void *)(ip6 + 1);
+		if ((void *)(udp +1) > data_end) {
+			return end(ctx, XDP_ABORTED, is_ipv6);
+		}
+		__u16 udp_len = data_end - (void *)udp;
+		udp->len = bpf_htons(udp_len);
+		udp->check = 0;
+		cs = pseudo_sum_ipv6(ip6, udp->len);
+	} else {
+		to_csum_len += sizeof(*xor);
+		udp = (void *)(ip + 1);
+		if ((void *)(udp +1) > data_end) {
+			return end(ctx, XDP_ABORTED, is_ipv6);
+		}
+		__u16 udp_len = data_end - (void *)udp;
+		udp->len = bpf_htons(udp_len);
+		udp->check = 0;
+		cs = pseudo_sum_ipv4(ip, udp->len);
+	}
+	if ((void *)udp + to_csum_len > data_end) {
+		return end(ctx, XDP_ABORTED, is_ipv6);
+	}
+	cs = bpf_csum_diff(0, 0, (void*)udp, to_csum_len, cs);
+	udp->check = csum_fold_flip(cs);
+	return end(ctx, XDP_TX, is_ipv6);
+}
```

Comment:
missing end of file newline

Reply:
updated in https://github.com/tailscale/tailscale/pull/12390/commits/b5b3393c9031da97bacc6b3c7c4410e93870f0ae